### DISCUSSION TITLE
`brew cleanup` deletes `r` site-library

### DISCUSSION THREAD

**Discussion by ywwry66** [Homebrew/discussions#4343] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.9-2-gaa36104
ORIGIN: https://github.com/Homebrew/brew
HEAD: aa36104e0f9785fd7d6e96bf312111cede3e94d0
Last commit: 3 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 2816c33e566b7d3f566d8c3cbb7968e68a2e9cbd
Core tap last commit: 31 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.mqFBIUt0bq/org.xquartz:0
HOMEBREW_EDITOR: emacs
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_NO_INSTALL_FROM_API: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.40.0 => /opt/homebrew/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2.1-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

Issue: Running `brew cleanup` right after a brand new `r` installation will unintentionally delete `r` site-library folder. To reproduce:
1. delete old `r` installation **(BACK UP YOUR R PACKAGES BEFORE RUNING THE CODE!)**
```sh
brew uninstall r
rm -r /opt/homebrew/lib/R
```
2. install `r`
```sh
brew install r
```
This will create site-library folder `/opt/homebrew/lib/R/4.2/site-library`
3.
```sh
brew cleanup
```
This immediately deletes `/opt/homebrew/lib/R/4.2`

**Comment by carlocab** [comment]
Can you see if this patch helps?
```diff
diff --git a/Formula/r.rb b/Formula/r.rb
index 48e4ffcc7f7..7667618347e 100644
--- a/Formula/r.rb
+++ b/Formula/r.rb
@@ -133,10 +133,10 @@ class R < Formula
   end
 
   def post_install
-    short_version =
-      `#{bin}/Rscript -e 'cat(as.character(getRversion()[1,1:2]))'`.strip
-    site_library = HOMEBREW_PREFIX/"lib/R/#{short_version}/site-library"
+    short_version = Utils.safe_popen_read(bin/"Rscript", "-e", "cat(as.character(getRversion()[1,1:2]))")
+    site_library = HOMEBREW_PREFIX/"lib/R"/short_version/"site-library"
     site_library.mkpath
+    touch site_library/".keepme"
     ln_s site_library, lib/"R/site-library"
   end
```

**Reply by ywwry66** [reply]
This solves the issue, thanks a lot!

However, when I was testing the change, I found that doing `brew postinstall r` may give incorrect behavior because of the `ln_s` line. The reason is `lib/"R/site-library"` is already a symlink to `site_library`, so doing `ln_s` again will follow the symlink and create a new link `site-library/site-library  ->  site-library`. To avoid this, `ln -snf` is the correct command in bash, but I did find an equivalent in `ruby`. I am not sure if this needs to be addressed since `brew install r` works fine, and I guess people rarely use `postinstall` command.

**Reply by ywwry66** [reply]
Looks like the approach in `python@3.11` works:
```diff
diff --git a/Formula/r.rb b/Formula/r.rb
index a1181fa5778..53a7478042f 100644
--- a/Formula/r.rb
+++ b/Formula/r.rb
@@ -142,7 +142,9 @@ class R < Formula
     site_library = HOMEBREW_PREFIX/"lib/R"/short_version/"site-library"
     site_library.mkpath
     touch site_library/".keepme"
-    ln_s site_library, lib/"R/site-library"
+    site_library_cellar = lib/"R/site-library"
+    site_library_cellar.unlink if site_library_cellar.exist?
+    site_library_cellar.parent.install_symlink site_library
   end

```
As a bonus, this creates a relative symlink.

**Comment by carlocab** [comment, marked as answer]
Handling this at Homebrew/homebrew-core#126232.

Do you have a log in at R's Bugzilla? Their `configure` script wrongly rejects curl 8+, which was released yesterday: https://daniel.haxx.se/blog/2023/03/20/curl-8-0-0-is-here/

As Daniel mentions, there is no API/ABI break in curl 8, so `configure` should be fine with it, but right now it rejects any curl with `LIBCURL_VERSION_MAJOR > 7`.

I'd report this myself, but the registering a new account is rather difficult.

**Reply by ywwry66** [reply]
Thanks for bringing this up. Unfortunately, I don't have a Bugzilla account either. They should be able to approve your application quickly as a Homebrew maintainer.
